usr/lib/icsf_stdll/icsf_specific.c

```diff
@@ -382,6 +384,179 @@ CK_RV icsftok_init(STDLL_TokData_t * tokdata, CK_SLOT_ID slot_id,
     return rc;
 }
 
+static void config_parse_error(int line, int col, const char *msg)
+{
+    TRACE_ERROR("Error parsing config file: line %d column %d: %s\n", line, col,
+                msg);
+}
+
+static struct icsf_config out_config;
+static char out_str_mech[64] = "";
+
+struct ref {
+    char *key;
+    char *addr;
+    size_t len;
+    int required;
+};
+
+static struct ref refs[] = {
+    { "token_name",        out_config.name,    sizeof(out_config.name),    1 },
+    { "token_manufacture", out_config.manuf,   sizeof(out_config.manuf),   1 },
+    { "token_model",       out_config.model,   sizeof(out_config.model),   1 },
+    { "token_serial",      out_config.serial,  sizeof(out_config.serial),  1 },
+    { "mech",              out_str_mech,       sizeof(out_str_mech),       1 },
+    { "uri",               out_config.uri,     sizeof(out_config.uri),     0 },
+    { "binddn",            out_config.dn,      sizeof(out_config.dn),      0 },
+    { "cacert",            out_config.ca_file, sizeof(out_config.ca_file), 0 },
+    { "cert",              out_config.cert_file, sizeof(out_config.cert_file), 0 },
+    { "key",               out_config.key_file, sizeof(out_config.key_file), 0 },
+};
+static const size_t refs_len = sizeof(refs)/sizeof(*refs);
+
+static int check_keys(const char *conf_name)
+{
+    size_t i;
+
+    for (i = 0; i < refs_len; i++) {
+        if (refs[i].required && *refs[i].addr == '\0') {
+            TRACE_ERROR("Missing required key \"%s\" in \"%s\".\n",
+                        refs[i].key, conf_name);
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static CK_RV config_parse_slot(const char *config_file,
+                               struct ConfigIdxStructNode *slot)
+{
+    struct ConfigBaseNode *c;
+    int i;
+    size_t k;
+    char *str;
+
+    TRACE_DEVEL("Slot: %lu\n", slot->idx);
+
+    confignode_foreach(c, slot->value, i) {
+        TRACE_DEVEL("Config node: '%s' type: %u line: %u\n",
+                    c->key, c->type, c->line);
+
+        str = confignode_getstr(c);
+        if (str != NULL) {
+            for (k = 0; k < refs_len; k++) {
+                if (!strcasecmp(refs[k].key, c->key)) {
+                    strncpy(refs[k].addr, str, refs[i].len);
+                    refs[k].addr[refs[k].len - 1] = '\0';
+                    goto found;
```

Comment:
Why not simply labeling the `confignode_foreach` and `continue <label>`?

Reply:
ah yes... sometimes things can be so much easier :-)

Reply:
ha, well, not really so easy....
C does not have a labeled loop/continue.
So I guess I have to leave it as is.